alerting/src/main/kotlin/com/amazon/opendistroforelasticsearch/alerting/model/AggregationTriggerRunResult.kt

```diff
@@ -0,0 +1,69 @@
+/*
+ *   Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License").
+ *   You may not use this file except in compliance with the License.
+ *   A copy of the License is located at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   or in the "license" file accompanying this file. This file is distributed
+ *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *   express or implied. See the License for the specific language governing
+ *   permissions and limitations under the License.
+ */
+package com.amazon.opendistroforelasticsearch.alerting.model
+
+import org.elasticsearch.common.io.stream.StreamInput
+import org.elasticsearch.common.io.stream.StreamOutput
+import org.elasticsearch.common.xcontent.ToXContent
+import org.elasticsearch.common.xcontent.XContentBuilder
+import java.io.IOException
+
+data class AggregationTriggerRunResult(
+    override var triggerName: String,
+    override var error: Exception? = null,
+    var aggregationResultBuckets: Map<String?, AggregationResultBucket>,
+    var actionResultsMap: MutableMap<String?, MutableMap<String, ActionRunResult>> = mutableMapOf()
+) : TriggerRunResult(triggerName, error) {
+
+    @Throws(IOException::class)
+    constructor(sin: StreamInput): this(
+        sin.readString(),
+        sin.readException() as Exception?, // error
+        sin.readMap(StreamInput::readString, ::AggregationResultBucket),
+        sin.readMap(StreamInput::readString) { innerSin: StreamInput -> innerSin.readMap(StreamInput::readString, ::ActionRunResult) }
+    )
+
+    override fun internalXContent(builder: XContentBuilder, params: ToXContent.Params): XContentBuilder {
+        return builder
+            .field(AGG_RESULT_BUCKETS, aggregationResultBuckets)
+            .field(ACTIONS_RESULTS, actionResultsMap as Map<String?, Any>?)
+    }
+
+    @Throws(IOException::class)
+    override fun writeTo(out: StreamOutput) {
+        super.writeTo(out)
+        out.writeMap<String?, AggregationResultBucket>(aggregationResultBuckets, StreamOutput::writeString) {
+                valueOut: StreamOutput, aggResultBucket: AggregationResultBucket -> aggResultBucket.writeTo(valueOut)
+        }
+
+        out.writeMap<String?, MutableMap<String, ActionRunResult>>(actionResultsMap, StreamOutput::writeString) {
+                actionOut: StreamOutput, value: MutableMap<String, ActionRunResult> -> actionOut.writeMap<String, ActionRunResult>(
+            value, StreamOutput::writeString) {
+                    actionResultOut: StreamOutput, actionResult: ActionRunResult -> actionResult.writeTo(actionResultOut)
+            }
+        }
```

Comment:
I think you might be able to simplify these to just `out.writeMap()` since the values of the Map are all writeables. It should call the Writer interface's implementation of write for us.

Could you try to make that change locally and use the round-trip Writeable unit tests to verify?

Reply:
I think i tried the `writeMap()` and if you'll see the implementation, it tries to write it as a generic value and would fail given the value here is not of a generic type or its `Writable`. I might be missing something, so will check again.